### ISSUE TITLE
go: not possible to detect reference tags using just kinds

### ISSUE THREAD

**Issue by techee** [universal-ctags/ctags#3211]
(This is a continuation of discussion from https://github.com/geany/geany/pull/3032)

For some parsers, such as go, some kinds are defined with `.referenceOnly = false` such as:
```
{true, 'p', "package", "packages",
	  .referenceOnly = false, ATTACH_ROLES (GoPackageRoles)},
```
These kinds serve both for indicating definition and reference tags.

The problem is, that to determine whether a tag is a definition tag or a reference tag, the kind letter isn't sufficient. For instance, when
```
package main

import "fmt"

func main() {
    fmt.Println("Hello, World!")
}
```
is processed with `ctags --extras=+r hello_world.go`, one gets
```
fmt	ctags/test.go	/^import "fmt"$/;"	p
main	ctags/test.go	/^func main() {$/;"	f	package:main
main	ctags/test.go	/^package main$/;"	p
```
Notice that the `p` tags `main` and `fmt` are indistinguishable in such output and look like identical tags even though `main` is a package definition and `fmt` is a package import (and a reference tag). I'm sure that when one passes the right command-line options to ctags, the role of the tag will become clear but IMO such fundamentally different things should be visible directly from the kind letter.

In other words, I believe (and I may be wrong of course) that there should be no `.referenceOnly = false` kind definitions and there should always be separate kinds for reference tags and definition tags.

**Comment by techee** [comment]
To clarify what this means for Geany - we currently enable all kinds (we don't do anything about roles but I assume some of them are enabled automatically by default). For instance, we need the `p` role for the package definition in go and currently, as a bonus, we also get the reference tags that we cannot distinguish from the definition tags. We could:

1. Check roleBits to see if a tag is a reference tag or a definition tag
2. Disable all roles for languages where `.referenceOnly = false` is used

I'd prefer avoiding (1) because it's not clear what bit means what and we'd have to somehow match it against reference definitions so I'd rather do (2) if the current behavior of ctags is intended.

**Comment by masatake** [comment]
Reference tags are the most advanced feature and unexplored area.
So I expect a user knows what happens with the option.

```c
0: size_t my_strlen(char *s)
1: {
2:    if (*s) return 1 + my_strlen(s+1);
3:    else    return 0;
4:}
...
n:    register_function("strlen", my_strlen);
```
The C parser assigns `f/function` to the `my_strlen` at line 0.
What kind will you give to `my_strlen` in line 3? How about line n?
I think they should have a same kind, "function".

The C parser assigns `z/parameter` to the `s` at line 0.
What kind will you give to the first `s` and the second `s` ?

I can change the questions. What kind of output an ideal ctags makes for the input?
```
my_strlen	input.c	/^size_t my_strlen(char *s)$/;"	f	roles:def
s	input.c	/^size_t my_strlen(char *s)$/;"	z	scope:function:my_strlen	roles:def
s	input.c	/^    if (*s) return 1 + my_strlen(s+1);$/;"	z	scope:function:my_strlen	roles:dereferenced
my_strlen	input.c	/^    if (*s) return 1 + my_strlen(s+1);$/;"	f	roles:called
s	input.c	/^    if (*s) return 1 + my_strlen(s+1);$/;"	z	scope:function:my_strlen	roles:lvalue
my_strlen	input.c	/^register_function("strlen", my_strlen);$/;"	f	roles:arg
```
I wonder this is overkill. Whether this is overkill or not is up to the goal of a client tool.

Do you really want to assign different kind letters to them instead of introducing roles?
```
$ ./ctags --list-kinds=C++ | wc -l
22
```
A kind namespace is very narrow; you can define only `(- (* 26 2) 1)` => 51.

To distinguish the difference in how a language object is referenced, I needed wider namespaces.
So I designed the concept, "roles".

I think you are not interested in roles. The difference between "lvalue" or "dereferenced" may not be important at all.
Not passing `--extras+r` is one of choice.
If you just want to know whether a tag is a reference tag or a definition tag, `--extras=+E` is for you.

```
$ ./ctags --extras=+r --fields=+E -o - /tmp/input.go
fmt	/tmp/input.go	/^import "fmt"$/;"	p	extras:reference
main	/tmp/input.go	/^func main() {$/;"	f	package:main
main	/tmp/input.go	/^package main$/;"	p
$ ./ctags --extras=+r --fields=+Er -o - /tmp/input.go
fmt	/tmp/input.go	/^import "fmt"$/;"	p	roles:imported	extras:reference
main	/tmp/input.go	/^func main() {$/;"	f	package:main	roles:def
main	/tmp/input.go	/^package main$/;"	p	roles:def
```

The `--extras=+E` option attaches `extras:(.*,)?reference(,.*)` to reference tags.

gtags is a tool that wants ctags to collect reference tags but doesn't want to know their roles.
gtags may use the following command line to call ctags:
```
$ ./ctags -x --_xformat="%R %-16N %4n %-16F %C" --extra=+r --fields=+r /tmp/input.go
./ctags -x --_xformat="%R %-16N %4n %-16F %C" --extra=+r --fields=+r /tmp/input.go
ctags: Warning: --extra option is obsolete; use --extras instead
D main                1 /tmp/input.go    package main
D main                5 /tmp/input.go    func main() {
R fmt                 3 /tmp/input.go    import "fmt"
```

I studied the concept, "roles" in #569. It is not closed yet.

**Comment by masatake** [comment]
@techee, I had a fundamental question; does geany need reference tags?
As far as reading your comment, it doesn't. What you need is just specifying `--extras=-r`, I think.
Am I wrong?

**Comment by techee** [comment]
> @techee, I had a fundamental question; does geany need reference tags?
> As far as reading your comment, it doesn't. What you need is just specifying --extras=-r, I think.
> Am I wrong?

They may be useful in some cases, check https://github.com/geany/geany/pull/3031.

> What you need is just specifying --extras=-r, I think.

We'll just have to do it in the code and I'll just do it for the languages where `.referenceOnly = false` is used as reference tags may be useful or us. Is something like this the correct approach?
```
static void disable_all_roles(const TMParserType lang, int kind)
{
	unsigned int c = countLanguageRoles(lang, kind);

	for (unsigned int i = 0; i < c; i++)
	{
		roleDefinition* rdef = getLanguageRole(lang, kind, (int)i);
		enableRole(rdef, false);
	}
}
```

**Comment by masatake** [comment]
I read geany/geany#303. Amazing.
I had taken a long time for designing the kind/role combination for Python to allow a client tool to provide such a feature!

> We'll just have to do it in the code and I'll just do it for the languages where .referenceOnly = false is used as reference tags may be useful or us. Is something like this the correct approach?

Definitely yes. I assumed after turning off all roles, geany turns on kinds what it really needs. 
If you will use such code, please, consider updating our mini-geany. 

If you cannot find a good function for your purpose, make an issue.

**Comment by techee** [comment]
> I read geany/geany#303. Amazing.

Is the cython problem from geany/geany#3031 something I should create an issue for in uctags? I'm not sure if cython is officially supported or not.

> I had taken a long time for designing the kind/role combination for Python to allow a client tool to provide such a feature!

Great job, we'll pretend it was made by us in the next Geany release ;-)

> Definitely yes. I assumed after turning off all roles, geany turns on kinds what it really needs.

We actually enable all kinds for all languages in ctags and then in our code decide whether we are interested in the kind or not. We don't do anything about roles right now, they just happen to be enabled for some kinds by default.

> If you will use such code, please, consider updating our mini-geany.

Definitely will.

I'm closing this issue as it appears to be the intended behavior.

**Comment by masatake** [comment]
> Is the cython problem from geany/geany#3031 something I should create an issue for in uctags? I'm not sure if cython is officially supported or not.

Sure. A good bug report drives the development. I don't promise I will work on it but opening an issue is welcome.
u-ctags will support all the languages as far as it uses only ascii-code:-)

**Comment by masatake** [comment]
I just see the screenshot of Geany.
I don't know how it is implemented but I guess Geany's internal tag database deals only with kinds; there is no area to record roles.

If the trouble comes from this limitation, how about embedding/encoding the roles information to the kind slots?

I expect the kind slot in Geany's db is represented in char or int.
Though a kind letter is represented in char, ctags only uses `[a-zA-Z]`.
You can map ctags's "p" kind + "imported" role to `(char) ('z' + 1)` in Go lang. If you use `int` for it. You can do as you want:-P.

I don't know how the label like "package" or "packages" in the symbol tree in the side bar is defined. But "package (imported)" can be one of good candidates. It can be generated from `snprintf(label, "%s (%s)", kind_name, role_name)`.

Just an idea.

**Comment by techee** [comment]
> I just see the screenshot of Geany.
> I don't know how it is implemented but I guess Geany's internal tag database deals only with kinds; there is no area to record roles.

We don't actually store kinds themselves but elements of an enum to which kinds are mapped here for individual languages:

https://github.com/geany/geany/blob/master/src/tagmanager/tm_parser.c

This mapping currently takes only kinds into account and adding roles would just complicate things and isn't worth it IMO. I just pushed a patch to Geany that disables roles generation for the Go 'p' kind and it resolves the issue for us - Go seems to be the only problematic language for now.

Oh and by the way, there's no database - it all happens in the memory only. Basically Geany is realtime ctags. We have a global array of tags containing tags from all open documents sorted by name (or when you use my ProjectOrganizer plugin, all files in a given directory) and then per-file array of tags again sorted by name. Whenever you type something in the editor, the old tags for the given file are removed from the global array, the updated document is re-parsed, the new tags are stored into the per-file array, sorted, and then merged into the global tag array. The global tag array is for instance used for autocompletion where you need symbols from all documents. All this scales surprisingly well, I tried it with complete linux kernel sources with millions of entries in the global tag array and you can just type text normally in the editor without any delays (and all this runs on the main thread - we don't use background threads in Geany).
